### DISCUSSION TITLE
Is Flux already usable in comfy?

### DISCUSSION THREAD

**Discussion by BrechtCorbeel** [Comfy-Org/ComfyUI#4173] [General]
Anyone tried it yet?

**Comment by CHNtentes** [comment]
4070 12G, use both t5 text encoder and flux dev unet in fp8 mode, takes about 2 min for each pic

**Comment by ltdrdata** [comment]
https://github.com/comfyanonymous/ComfyUI_examples/tree/master/flux

This feature is supported from day zero.

**Comment by tomscwills** [comment]
Using the base Flux1_dev workflow taxes my setup,
64g ram, 4090, using t5xxl_fp16. 100% Vram usage.
Will say img output is good though.
![image](https://github.com/user-attachments/assets/59e07d1b-f24c-4ea2-8990-e38f1daeb3f6)

**Comment by JorgeR81** [comment]
On the other end of the hardware spectrum ...

Intel i7-7700
32 GB RAM
GTX 1070 (8GB VRAM)
Windows 10

Thanks to the "schnell" version, it's usable for me. 
This is the example workflow, ( with 1024 x 1024 image size ).
But, I'm using the [ t5xxl_fp8_e4m3fn ] text encoder, instead.

![flux](https://github.com/user-attachments/assets/d5d45635-d15c-4a36-b703-24264541b0b0)

**Comment by JorgeR81** [comment]
I can also use the "dev" model, with both the model and text encoder in FP16 !

The "dev" has the same times per iteration, but it requires more steps.
With text encoder at F16, times are the same.
And with the model at FP16, it's just a little slower. 

It takes a long time, but it's really nice to be able to get the maximum quality !
The preview works, so I can cancel, if I don't like the seed ... 

![flux2](https://github.com/user-attachments/assets/19d3555d-8257-4722-9b9f-d8e87046794f)

**Comment by tpcdaz** [comment]
I have an RTX 3060 12gb edition and the schnell model works lovely.

**Reply by a3nima** [reply]
doesnt it go into lowvram mode? check the command line prompt window.

**Comment by BrechtCorbeel** [comment]
I seem to get no step closer to using it error after error with comfy right now

**Comment by AngryLoki** [comment]
For those who are interested,FLUX.1-schnell fp16 on modern CPU (Ryzen 7950x3d) is not massively slower than low-end GPU:
```
100%|██████████████████████████████████████████████| 4/4 [03:14<00:00, 48.66s/it]
Using split attention in VAE
Using split attention in VAE
Requested to load AutoencodingEngine
Loading 1 new model
Prompt executed in 242.54 seconds
```
but consumes up to 54GB RAM. Probably it casts all models to fp32 (adding `--bf16-vae --bf16-unet` changes nothing).

**Reply by BrechtCorbeel** [reply]
> For those who are interested,FLUX.1-schnell fp16 on modern CPU (Ryzen 7950x3d) is not massively slower than low-end GPU:
> 
> ```
> 100%|██████████████████████████████████████████████| 4/4 [03:14<00:00, 48.66s/it]
> Using split attention in VAE
> Using split attention in VAE
> Requested to load AutoencodingEngine
> Loading 1 new model
> Prompt executed in 242.54 seconds
> ```
> 
> but consumes up to 54GB RAM. Probably it casts all models to fp32 (adding `--bf16-vae --bf16-unet` changes nothing).

what is the size of your image at the point that this happens?

I just max out the ram on any pc one of mine has 96GB the new one 128GB, because everything adds up so quick these days.

Don't expect to be running AI on 300$ laptops, the big companies use 80GB VRAM GPUS with TBs of RAM to run their largest AI models.

**Reply by AngryLoki** [reply]
> what is the size of your image at the point that this happens?

1024x1024, workflow copied from https://comfyanonymous.github.io/ComfyUI_examples/flux/#flux-schnell-1

Actually, I slightly edited https://github.com/comfyanonymous/ComfyUI/pull/3649 and with bf16 datatype (again on CPU) unet is 2 times faster:

```
100%|██████████████████████████████████████████████| 4/4 [01:40<00:00, 25.18s/it]
Using split attention in VAE
Using split attention in VAE
Requested to load AutoencodingEngine
Loading 1 new model
Prompt executed in 118.42 seconds
```

And memory consumption is 35GB, which approximately equals 23G of flux1-schnell.sft (it is already fully bfloat16) + 9.2G of t5xxl_fp16 (it should be converted from fp16 to bf16) + ae + clip_l + internal structures.

**Reply by BrechtCorbeel** [reply]
> > what is the size of your image at the point that this happens?
> 
> 1024x1024, workflow copied from https://comfyanonymous.github.io/ComfyUI_examples/flux/#flux-schnell-1
> 
> Actually, I slightly edited #3649 and with bf16 datatype (again on CPU) unet is 2 times faster:
> 
> ```
> 100%|██████████████████████████████████████████████| 4/4 [01:40<00:00, 25.18s/it]
> Using split attention in VAE
> Using split attention in VAE
> Requested to load AutoencodingEngine
> Loading 1 new model
> Prompt executed in 118.42 seconds
> ```
> 
> And memory consumption is 35GB, which approximately equals 23G of flux1-schnell.sft (it is already fully bfloat16) + 9.2G of t5xxl_fp16 (it should be converted from fp16 to bf16) + ae + clip_l + internal structures.

Where does your concern come from exactly? you are using all the highest settings, from running it I see it loads itself into memory and stays there unlike stable diffusion, which unloaded an reloaded all the time.

Is it because you do not have much left and are using the computer?

**Reply by AngryLoki** [reply]
> Where does your concern come from exactly?

No concerns from my side, I'm just checking that ComfyUI does not allocate fp32 memory for bf16 model, causing OOM kills. 
Safetensor files are not compressed, so my expectation is that loaded model memory consumption is the same as file size.

**Reply by CHNtentes** [reply]
> > what is the size of your image at the point that this happens?
> 
> 1024x1024, workflow copied from https://comfyanonymous.github.io/ComfyUI_examples/flux/#flux-schnell-1
> 
> Actually, I slightly edited #3649 and with bf16 datatype (again on CPU) unet is 2 times faster:
> 
> ```
> 100%|██████████████████████████████████████████████| 4/4 [01:40<00:00, 25.18s/it]
> Using split attention in VAE
> Using split attention in VAE
> Requested to load AutoencodingEngine
> Loading 1 new model
> Prompt executed in 118.42 seconds
> ```
> 
> And memory consumption is 35GB, which approximately equals 23G of flux1-schnell.sft (it is already fully bfloat16) + 9.2G of t5xxl_fp16 (it should be converted from fp16 to bf16) + ae + clip_l + internal structures.

why cast fp16 to bf16? that will also cause loss, won't it?

**Reply by JorgeR81** [reply]
I've also been doing some testing on this, ( see this discussion: https://github.com/comfyanonymous/ComfyUI/discussions/4226 )
I have 32 GB RAM and 8 GB VRAM.

Even with the Comfy-org (fp8) checkpoint ( 17.2 GB ), with text encoders included, my RAM usage still reaches the limit, when loading the model, and the SSD activity spikes. 
It works, but it wears down the SSD, I guess ...

**Reply by AngryLoki** [reply]
> why cast fp16 to bf16? that will also cause loss, won't it?

Zen4 CPUs can't perform FP16 computations (other than convert to FP32).
Also e4m3fn has also no native support, so that's not an option. Yes, FP16->BF16 is a loss, but probably not as big as anything->e4m3fn. Ideally I need to take fp32 model and resave it as BF16, but I don't expect effect to be significant.

**Reply by RandomGitUser321** [reply]
Within the https://github.com/comfyanonymous/ComfyUI/blob/1c08bf35b49879115dedd8ec6bc92d9e8d8fd871/comfy/supported_models.py#L631

`supported_inference_dtypes = [torch.bfloat16, torch.float32]`
So it only supports them

**Comment by drulia** [comment]
ComfyUI noob here, I have downloaded fresh ComfyUI windows portable, downloaded t5xxl_fp16.safetensors and clip_l.safetensors and vae to run 
FLUX.1-schnell. I'm running it using RTX 4070 Ti SUPER and system has 128GB of ram. But it takes 670 seconds to render one example image of galaxy in a bottle. https://comfyanonymous.github.io/ComfyUI_examples/flux/

Repeated runs take just as long. Running on CPU (i7-14700KF) is pretty much same story. Is this expected or I'm doing something wrong?

![Screenshot 2024-08-06 223446](https://github.com/user-attachments/assets/ee126a62-4b84-4200-aa34-2a7f9d812c9d)

**Reply by JorgeR81** [reply]
There is definitely something wrong. 
You should probably open an issue to get feedback from the developers.

**Reply by askmyteapot** [reply]
for a 4070ti super, you will need to downcast to fp8_e5m2 (or fp8_e4m3fn) otherwise it will be painfully slow. My guess is that you have sys ram fallback enabled too, which will slow things down even more as you exhaust your GPU Vram.

**Reply by JorgeR81** [reply]
Yeah but, there must be something else going wrong, for @drulia.
I have an older GTX 1070 (8 GB VRAM). 
For me, using the fp16 model, it's slower, but only slightly. 

I only have 32 GB RAM, so during model loading, I run out of RAM and the system needs to use the the page file. 
https://github.com/comfyanonymous/ComfyUI/discussions/4226#discussioncomment-10259959

But, I can still generate over 4 times faster than that. 
https://github.com/comfyanonymous/ComfyUI/discussions/4173#discussioncomment-10229219

**Comment by DavidSH92** [comment]
I want to test the fp8 model locally,but an error was reported ,how do I install the FluxGuidance node，comfui managr cannot find this node？

**Reply by JorgeR81** [reply]
The FluxGuidance is not a custom node. 
It should be in Comfy UI, by default, if it's updated.

**Reply by DavidSH92** [reply]
Wow~ ⊙o⊙，yesterday, I updated Comfui through ComfyUI manager and it prompted 'It has been successfully updated.' I was unable to run ComfyUI_examples successfully, but today I started Comfui and no longer reported FluxGuidance related errors. Perhaps I need to restart my computer? I'm not quite sure, thank you for your answer

**Comment by a3nima** [comment]
```
got prompt
[rgthree] Using rgthree's optimized recursive execution.
Requested to load FluxClipModel_
Loading 1 new model
Requested to load Flux
Loading 1 new model
loading in lowvram mode 21973.2
100%|██████████████████████████████████████████████████████████████████████████████████| 25/25 [00:39<00:00, 1.59s/it] 
```

can I prevent this lowvram mode (24GB VRAM) ? or doesnt it hurt? (rest of system: 64GB Sys RAM, Windows 10)

**Reply by JorgeR81** [reply]
To prevent lowvram you can try the ( smaller ) fp8 versions of the Flux models and t5 encoder ( at a slight quality loss ).
https://comfyanonymous.github.io/ComfyUI_examples/flux/
https://huggingface.co/Kijai/flux-fp8/tree/main

But you're looking pretty good ! 
Much faster than other 4090 here: https://github.com/comfyanonymous/ComfyUI/discussions/4173#discussioncomment-10225864

**Reply by a3nima** [reply]
thanks, Jorge! PS: 3090 to be precise :D I'm only using the workflow provided by comfyUI and updated everything incl. python before using flux.

**Reply by JorgeR81** [reply]
Hey, this user has the same GPU, and less RAM, and doesn't seem to need lowvram.
https://github.com/comfyanonymous/ComfyUI/issues/4271#issuecomment-2276812629

**Reply by a3nima** [reply]
Maybe it has to do with the comfyUI version? I'm running windows standalone portable.

**Reply by JorgeR81** [reply]
I think he is using a specific  `--highvram` cmd
You can add them, by editing your [ run_nvidia_gpu.bat ] file ( the one you click on to start Comfy UI )
https://github.com/comfyanonymous/ComfyUI/discussions/1043#discussioncomment-6590959
https://github.com/comfyanonymous/ComfyUI/issues/4271#issuecomment-2276643565

**Comment by JorgeR81** [comment]
Do you have slower generation times after the latest update ?
https://github.com/comfyanonymous/ComfyUI/issues/4271

**Comment by JorgeR81** [comment]
Flux seems good at refining its own images.
With other models, the images tend to get burned, especially if the same seed is used. 

![flux](https://github.com/user-attachments/assets/f0d88cab-bfa3-4f25-abe7-cf1d08df2f51)

In this example, the skin texture looks better with 12 + 4 steps more than 25 steps. 

**_12 + 4 steps_** ( workflow in the png )

![12_4](https://github.com/user-attachments/assets/10973245-ee32-4920-b360-f26e7271a81e)

_**25 steps**_

![25](https://github.com/user-attachments/assets/c269e539-8703-4adf-bf7a-03c061001072)

**Reply by a3nima** [reply]
doesnt this also speak for latent upscale (hiresfix) instead of using upscalers? I'm new at comfyUI, but I think I would have a workflow that has a switch like in A1111, so that I can choose to upscale with either upscaler or latent. (normally I prefer latent)

**Reply by JorgeR81** [reply]
Yes, you can add an [ Upscale Latent By ] node to upscale the latent image, between the two KSamplers ( pink wires ).
But you can also add an [ Upscale Image By ] node to do a regular upscale, after the image was generated ( blue wires ).

But you need to manually connect the nodes. 
Comfy UI doesn't have "switch" nodes by default.
For that you need a custom node pack, with extra nodes, like this one. 
https://github.com/ltdrdata/ComfyUI-Impact-Pack

You can install multiple custom node packs more easily with the Comfy UI Manager.
https://github.com/ltdrdata/ComfyUI-Manager

**Reply by JorgeR81** [reply]
You may need to use more steps in the first image, to get a better quality starting point, for a latent or regular upscaling,
Or you can upscale the refined image instead.
